compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -0,0 +1,254 @@
+    forM_ existingChoices $ \choice -> do
+        withContext (ContextTemplate (present module_) (present template) (TPChoice (present choice))) $ do
+            let returnTypesMatch = foldU alphaType (fmap chcReturnType choice)
+            unless returnTypesMatch $
+                throwWithContext (EUpgradeChoiceChangedReturnType (NM.name (present choice)))
+
+            throwIfDifferent "controllers" (extractFuncFromFuncThisArg . chcControllers <$> choice) $
+                EUpgradeChoiceChangedControllers $ NM.name $ present choice
+
+            let observersErr = EUpgradeChoiceChangedObservers $ NM.name $ present choice
+            case fmap (mapENilToNothing . chcObservers) choice of
+               Upgrading { past = Nothing, present = Nothing } -> do
+                   pure ()
+               Upgrading { past = Just past, present = Just present } -> do
+                   throwIfDifferent "observers" (extractFuncFromFuncThisArg <$> Upgrading past present) observersErr
+               _ -> do
+                   throwWithContext observersErr
+
+            let authorizersErr = EUpgradeChoiceChangedAuthorizers $ NM.name $ present choice
+            case fmap (mapENilToNothing . chcAuthorizers) choice of
+               Upgrading { past = Nothing, present = Nothing } -> pure ()
+               Upgrading { past = Just past, present = Just present } ->
+                   throwIfDifferent "authorizers" (extractFuncFromFuncThisArg <$> Upgrading past present) authorizersErr
+               _ -> throwWithContext authorizersErr
+        pure choice
+
+    -- This check assumes that we encode signatories etc. on a template as
+    -- $<uniquename> this, where $<uniquename> is a function that contains the
+    -- actual definition. We resolve this function and check that it is
+    -- identical.
+    withContext (ContextTemplate (present module_) (present template) TPPrecondition) $
+        throwIfDifferent "precondition" (extractFuncFromCaseFuncThis . tplPrecondition <$> template) $
+            EUpgradeTemplateChangedPrecondition $ NM.name $ present template
+    withContext (ContextTemplate (present module_) (present template) TPSignatories) $
+        throwIfDifferent "signatories" (extractFuncFromFuncThis . tplSignatories <$> template) $
+            EUpgradeTemplateChangedSignatories $ NM.name $ present template
+    withContext (ContextTemplate (present module_) (present template) TPObservers) $
+        throwIfDifferent "observers" (extractFuncFromFuncThis . tplObservers <$> template) $
+            EUpgradeTemplateChangedObservers $ NM.name $ present template
+    withContext (ContextTemplate (present module_) (present template) TPAgreement) $
+        throwIfDifferent "agreement" (extractFuncFromFuncThis . tplAgreement <$> template) $
+            EUpgradeTemplateChangedAgreement $ NM.name $ present template
```

Comment:
I don't think we should be checking the template functions, since on the one hand a syntactic check would miss some problematic changes:

```haskell
template T with 
    p, q : Party
    ..
  where
    signatory sigHelper this -- accepted because syntactically it's just the same 
                             -- (package-)local function applied to the same variable

sigHelper : T -> [Party]
sigHelper T {..} =
  [p] -- v1
  [p, q] -- v2
```

while on the other, as pointed out by Andreas in the design doc, it would prevent legitimate changes:

```haskell
template T with
    p, q : Party
    r : Optional Party -- added in v2
  signatory [p, q] -- v1
  signatory [p, q] <> optionalToList r -- v2, rejected because it's syntactically different, 
                                       -- even though it agrees with v1 for upgraded contracts
```

Reply:
I'm working on changing these to warnings as per discussion with Bas, but it's a bit large of a change so I'm putting it into its own PR

Reply:
https://github.com/digital-asset/daml/pull/17609